### DISCUSSION TITLE
How incremental is Roslyn?

### DISCUSSION THREAD

**Discussion by viluon** [dotnet/roslyn#67696] [Q&A]
### Background

I'm trying to understand the Roslyn architecture w.r.t. incremental computation, change propagation, and state management. I had a quick look at the repository, but couldn't find any comprehensive internal architecture overview. The [MS docs](https://learn.microsoft.com/en-gb/dotnet/csharp/roslyn-sdk/) overall seem to take a very black-box view of the Roslyn compilers and don't discuss change processing at all. Feel free to simply answer with a link if there's a useful resource I missed. I'm building a language server for my master's thesis and I'm looking for inspiration and engineering ideas in the wild.

In a [talk about modern compiler design](https://www.youtube.com/watch?v=wSdV1M7n4gQ), Anders Hejlsberg described an incremental approach for updating syntax trees to provide quick feedback to the programmer during an editing session. I know about the concept of [red-green trees](https://ericlippert.com/2012/06/08/red-green-trees/) and I want to know whether there are ideas like it further down the pipeline. Anders mentioned that Roslyn is effectively a pull model which recomputes necessary information on-demand when requested. He hinted at the compiler simply requesting everything required for a full compilation run. I'm wondering how deep does this idea go.

### Questions

Is Roslyn a [query-based compiler](https://ollef.github.io/blog/posts/query-based-compilers.html)? (`rustc`, to an extent, [is](https://rustc-dev-guide.rust-lang.org/overview.html#queries).) Is there a systematic approach to incremental computation applied throughout the codebase, or is incrementality handcrafted only at key points? What about the lazy pull model? Is there a methodical approach to memoization and persistence? Expert opinion and resources would be very helpful.

**Comment by CyrusNajmabadi** [comment]
@viluon  i would recommend going to discord.gg/csharp and then the roslyn channel.  We can chat there, which is likely be a better way to handle these questions.

**Reply by viluon** [reply]
Thank you for the offer; I'll take what I can get! I think having this information in a place that makes it more discoverable and easier to search for than Discord would be more helpful. I could compile my findings into an answer and post it here later.
